File: src/main/java/org/modelingvalue/nelumbo/syntax/ParserResult.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.syntax;

import org.modelingvalue.collections.List;
import org.modelingvalue.collections.mutable.MutableList;
import org.modelingvalue.nelumbo.Evaluatable;
import org.modelingvalue.nelumbo.KnowledgeBase;
import org.modelingvalue.nelumbo.Node;
import org.modelingvalue.nelumbo.U;
import org.modelingvalue.nelumbo.collections.NList;
import org.modelingvalue.nelumbo.syntax.Tokenizer.TokenizerResult;

public class ParserResult implements ParseExceptionHandler {

    private final TokenizerResult             tokenizerResult;
    private final boolean                     throwing;
    private final MutableList<ParseException> exceptions;

    private Node                              root;

    public ParserResult(TokenizerResult tokenizerResult, boolean throwing) {
        this.tokenizerResult = tokenizerResult;
        this.throwing = throwing;
        this.exceptions = MutableList.of(List.of());
    }

    public List<Node> roots() {
        return root instanceof NList listRoot ? listRoot.elementsFlattened() : root != null ? List.of(root) : List.of();
    }

    public Node root() {
        return root;
    }

    public void setRoot(Node root) {
        this.root = root;
    }

    @Override
    public void addException(ParseException exception) throws ParseException {
        if (throwing) {
            throw exception;
        }
        this.exceptions.add(exception);
    }

    @Override
    public List<ParseException> exceptions() {
        return exceptions.toImmutable();
    }

    public void throwException() throws ParseException {
        if (!exceptions.isEmpty()) {
            throw exceptions().first();
        }
    }

    public void evaluate() throws ParseException {
        if (exceptions.isEmpty()) {
            KnowledgeBase knowledgeBase = KnowledgeBase.CURRENT.get();
            for (Node root : roots()) {
                if (root instanceof Evaluatable eval) {
                    eval.evaluate(knowledgeBase, this);
                }
            }
        }
    }

    public void print() {
        for (ParseException exc : exceptions()) {
            System.out.println(exc);
        }
        for (Node root : roots()) {
            System.out.println(root);
        }
    }

    public void checkAssertions() {
        if (false && tokenizerResult != null && root != null && U.areAssertsEnabled()) {
            StringBuffer sb = new StringBuffer();
            root.deparse(sb);
            String result = sb.toString();
            String input = tokenizerResult.input();
            assert input.equals(result);
        }
    }

}

File: web-ts/src/syntax/ParserResult.ts
/**
 * ParserResult - full parse result with error collection.
 * @JAVA_REF org.modelingvalue.nelumbo.syntax.ParserResult
 */

import { List } from 'immutable';
import type { Node } from '../Node';
import type { TokenizerResult } from './Tokenizer';
import { ParseException } from './ParseException';

/**
 * ParserResult - result of parsing.
 */
export class ParserResult {
  private readonly _tokenizerResult: TokenizerResult;
  private readonly _throwing: boolean;

  private _exceptions: ParseException[] = [];
  private _root: Node | null = null;

  constructor(tokenizerResult: TokenizerResult, throwing: boolean) {
    this._tokenizerResult = tokenizerResult;
    this._throwing = throwing;
  }

  /**
   * Get the tokenizer result.
   */
  tokenizerResult(): TokenizerResult {
    return this._tokenizerResult;
  }

  /**
   * Get all root nodes.
   */
  roots(): List<Node> {
    // Check if root is a list node
    if (this._root !== null && this.isListNode(this._root)) {
      return (this._root as unknown as { elementsFlattened(): List<Node> }).elementsFlattened();
    }
    return this._root !== null ? List([this._root]) : List();
  }

  private isListNode(node: Node): boolean {
    return 'elementsFlattened' in node && typeof (node as unknown as { elementsFlattened: unknown }).elementsFlattened === 'function';
  }

  /**
   * Get the single root node.
   */
  root(): Node | null {
    return this._root;
  }

  /**
   * Set the root node.
   */
  setRoot(root: Node): void {
    this._root = root;
  }

  /**
   * Add an exception.
   */
  addException(exception: ParseException): void {
    if (this._throwing) {
      throw exception;
    }
    this._exceptions.push(exception);
  }

  /**
   * Get all exceptions.
   */
  exceptions(): List<ParseException> {
    return List(this._exceptions);
  }

  /**
   * Throw the first exception if any.
   */
  throwException(): void {
    if (this._exceptions.length > 0) {
      throw this._exceptions[0];
    }
  }

  /**
   * Evaluate all root nodes.
   */
  evaluate(): void {
    if (this._exceptions.length === 0) {
      for (const root of this.roots()) {
        if (this.isEvaluatable(root)) {
          // root.evaluate(knowledgeBase, this);
        }
      }
    }
  }

  private isEvaluatable(node: Node): boolean {
    return 'evaluate' in node && typeof (node as unknown as { evaluate: unknown }).evaluate === 'function';
  }

  /**
   * Print results to console.
   */
  print(): void {
    for (const exc of this._exceptions) {
      console.log(exc.fullMessage);
    }
    for (const root of this.roots()) {
      console.log(root.toString());
    }
  }

  /**
   * Check assertions (debug only).
   */
  checkAssertions(): void {
    // Assertions disabled in production
  }
}
